Add a global Twig variable with Contao state

aschempp opened the pull request:
I think we controversely discussed this before. However, I had to copy-paste this to multiple of my projects recently, and I think it would make sense in general. With this, you can do the following in any Twig template:

**any property from the page object (from current request = usually the global)**
```twig
{{ contao.page.urlPrefix }}
```

**check if preview mode is enabled**
```
{{ contao.tokenChecker.isPreviewMode }}
```

aschempp committed:
Add a global Twig variable with Contao state

ausi commented:
I think the downside of something like this is, that the controller rendering the template no longer knows what data the template accesses, making the response of that controller theoretically uncacheable.

However, that problem already exists because of the [global `app` variable](https://symfony.com/doc/current/templates.html#the-app-global-variable).

Maybe we could at least mention that in the documentation of this global twig variable then?

m-vo commented:
We also should discuss if this would better be a runtime that provides dedicated functions instead of a 'global'. I'm not sure yet but I'm leaning a bit towards the latter because then we could provide some more convenience instead of outputting raw values that the template has to deal with. 

@aschempp Out of interest: where did you need those things? I assume in content elements where you did not have control over the controllers? If so, we could also just add these things to `AbstractContentElementController#addDefaultDataToTemplate()`.

aschempp commented:
I mostly used them in `fe_page.html.twig` actually. Only output the analytics scripts if no backend user is logged in (and not in preview mode). And I think I added some custom fields to `tl_page` and added CSS classes based on that to the body.

In my cases, neither of these would affect the cache (since anything in regards to user being logged in means a session, and a session means no cache). And any page property wouldn't change on a per-user basis.

fritzmg commented:
At least the `contao.tokenChecker` would be required if you'd want to implement Analytics templates via Twig - since you'd want to use the `hasBackendUser` method of the token checker.

aschempp commented:
> At least the `contao.tokenChecker` would be required if you'd want to implement Analytics templates via Twig - since you'd want to use the `hasBackendUser` method of the token checker.

It would probably be better to add a `contao.hasBackendUser` method directly and use the token checker there. Then Twig users would not rely on _knowing_ there is a `TokenChecker` and what methods it contains.
But I don't know if it is possible (and how to) maybe have `contao.security.hasBackendUser` without populating the `contao.security` array with properties that are possibly never used.

andyKempf commented:
Servus & Moin!
Are there plans to include this change in the new LTS version? That would be nice, since you can't access tl_page DCA customizations with php variable $[GLOBALS] in Twig.

Thanks for your great work!

Greetings from Munich,
Andi

fritzmg commented:
I agree that we should add this (or any other global variables that are useful) to the new LTS. Could also be done as a bugfix.

ausi commented on core-bundle/src/Twig/Extension/ContaoExtension.php:
```diff
@@ -90,6 +93,11 @@ public function __toString(): string
 
+    public function getGlobals(): array
+    {
+        return [$this->contaoVariable];
```
```suggestion
        return ['contao' => $this->contaoVariable];
```

aschempp committed:
Merge remote-tracking branch 'upstream/5.3' into feature/contao-variable

# Conflicts:
#	core-bundle/config/services.yaml
#	core-bundle/src/Twig/Extension/ContaoExtension.php

aschempp committed:
Fixed class registration and added more properties

aschempp commented:
PR updated. I also think we should "group" more useful global properties into the `contao.` variables. That includes the request token to me. @contao/developers let me know if you agree, maybe I'll need to update all usages then.

I checked the `TemplateTrait` and a few other places but couldn't find anything else to add. But I'm sure we'll come up with more … once we start using it! I intentionally did not add `TL_MODE` because the scope should not be relevant to the templates (and already is available for fragments etc.).

aschempp committed:
CS

aschempp commented:
Also, the `SecurityVariable` is necessary to get autocomplete support in the IDE. I tried using an anonymous class first, but then PHPStorm would not autocomplete the properties.

aschempp committed:
Tests

ausi commented on core-bundle/src/Twig/Global/ContaoVariable.php:
```diff
@@ -0,0 +1,59 @@
+        return new SecurityVariable($this->tokenChecker);
+    }
+
+    public function requestToken(): string
```
```suggestion
    public function request_token(): string
```
Should we name this `request_token` to be compatible with naming conventions in Twig?

aschempp replied:
no, this is a function call, `app.request.requestUri` is uppercase as well. The usage in Twig would be `contao.requestToken` or `contao.security.previewMode` etc.

ausi replied:
```suggestion
    public function getRequest_token(): string
```
Looks like this in the naming convention to use here? see https://github.com/symfony/symfony/blob/d621936e8c638247de4c5d1cd75cc5ff55b84c3e/src/Symfony/Bridge/Twig/AppVariable.php#L196

fritzmg replied:
Yes, Symfony uses

* `app.current_route`
* `app.current_route_parameters`
* `app.enabled_locales`

https://symfony.com/doc/6.4/templates.html#the-app-global-variable

And our previous global variable was also snake case (`request_token`), so I think it should be `contao.request_token` 👍

aschempp replied:
that looks horrible to me …

aschempp committed:
CS

aschempp committed:
Tests

aschempp committed:
CS

ausi commented on core-bundle/config/services.yaml:
```diff
@@ -1150,6 +1150,7 @@ services:
             - '@twig'
             - '@contao.twig.filesystem_loader'
             - '@contao.csrf.token_manager'
+            - '@contao.twig.global.contao'
```
```suggestion
            - '@contao.twig.global.variable'
```

ausi commented on core-bundle/config/services.yaml:
```diff
@@ -1201,6 +1202,14 @@ services:
         arguments:
             - '@contao.framework'
 
+    contao.twig.global.contao:
```
```suggestion
    contao.twig.global.variable:
```

ausi replied:
see https://github.com/contao/contao/actions/runs/7671819427/job/20911013147?pr=6436

aschempp committed:
Fix properties

aschempp commented:
The properties are now 
```twig
{{ contao.has_backend_user }}
{{ contao.is_preview_mode }}
{{ contao.request_token }}
{{ contao.page.* }}
```

leofeyer commented on core-bundle/src/Twig/Global/ContaoVariable.php:
```diff
@@ -0,0 +1,64 @@
+            && (int) $GLOBALS['objPage']->id === (int) $pageModel
+        ) {
+            return $GLOBALS['objPage'];
+        }
```
This is legacy code from Contao 4.13. We have removed it everywhere in Contao 5, so it should be removed here as well, shouldn‘t it?

aschempp replied:
fixed in a0245b588f67b9227d898e81b36b7f3c4848b40d

ausi reviewed (dismissed):
LGTM! Apart from https://github.com/contao/contao/pull/6436#discussion_r1469255870

aschempp committed:
Simplify the pageModel getter

leofeyer committed:
CS

leofeyer commented:
Thank you @aschempp.

ausi commented on core-bundle/src/Twig/Global/ContaoVariable.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $pageModel = $this->requestStack->getCurrentRequest()?->attributes->get('pageModel');
+
+        if ($pageModel instanceof PageModel) {
+            return null;
+        }
+
+        return $pageModel;
```
This looks wrong to me.

```suggestion
        if (!$pageModel instanceof PageModel) {
            return null;
        }

        return $pageModel;
```

leofeyer replied:
Fixed in 2a215a11fd4e61b0351bd38efdcb4b674cf2e161.
